File: _archived/documents-nodejs/src/domain/entities/document.ts
import type { DocumentStatusValue, ScanStatusValue } from '@/shared/constants';

/**
 * Core Document entity — cloud and persistence agnostic.
 * No provider-specific logic allowed here.
 */
export interface Document {
  id:            string;
  tenantId:      string;
  productId:     string;
  referenceId:   string;          // Case ID, matter ID, patient ID, etc.
  referenceType: string;          // 'CASE' | 'PATIENT' | 'MATTER' | 'LIEN' | ...
  documentTypeId: string;
  title:         string;
  description:   string | null;
  status:        DocumentStatusValue;

  // Storage (never exposed to clients — internal only)
  storageKey:    string;
  storageBucket: string;
  mimeType:      string;
  fileSizeBytes: number;
  checksum:      string;          // SHA-256 of file content

  // Versioning
  currentVersionId: string | null;
  versionCount:  number;

  // Scan lifecycle (mirrors current version scan state on the document for fast access gating)
  scanStatus:        ScanStatusValue;
  scanCompletedAt:   Date | null;
  scanThreats:       string[];

  // Soft delete + retention
  isDeleted:     boolean;
  deletedAt:     Date | null;
  deletedBy:     string | null;
  retainUntil:   Date | null;
  legalHoldAt:   Date | null;

  // Audit
  createdAt:     Date;
  createdBy:     string;
  updatedAt:     Date;
  updatedBy:     string;
}

export interface DocumentType {
  id:          string;
  tenantId:    string | null;   // null = global type available to all tenants
  productId:   string | null;
  code:        string;
  label:       string;
  isActive:    boolean;
  createdAt:   Date;
}

/** Payload for creating a new document */
export interface CreateDocumentInput {
  tenantId:      string;
  productId:     string;
  referenceId:   string;
  referenceType: string;
  documentTypeId: string;
  title:         string;
  description?:  string;
  uploadedBy:    string;
  // Resolved during upload — not accepted from client
  storageKey?:   string;
  storageBucket?: string;
  mimeType?:     string;
  fileSizeBytes?: number;
  checksum?:     string;
}

/** Payload for updating document metadata */
export interface UpdateDocumentInput {
  title?:         string;
  description?:   string;
  documentTypeId?: string;
  status?:        DocumentStatusValue;
  retainUntil?:   Date;
  updatedBy:      string;
}

File: apps/services/documents/Documents.Domain/Entities/Document.cs
using Documents.Domain.Enums;

namespace Documents.Domain.Entities;

public sealed class Document
{
    public Guid   Id               { get; set; }
    public Guid   TenantId         { get; set; }
    public string ProductId        { get; set; } = string.Empty;
    public string ReferenceId      { get; set; } = string.Empty;
    public string ReferenceType    { get; set; } = string.Empty;
    public Guid   DocumentTypeId   { get; set; }
    public string Title            { get; set; } = string.Empty;
    public string? Description     { get; set; }
    public DocumentStatus Status   { get; set; } = DocumentStatus.Draft;
    public string MimeType         { get; set; } = string.Empty;
    public long   FileSizeBytes    { get; set; }

    // Storage — never exposed in API responses
    public string StorageKey       { get; set; } = string.Empty;
    public string StorageBucket    { get; set; } = string.Empty;
    public string? Checksum        { get; set; }

    // Versioning
    public Guid?  CurrentVersionId { get; set; }
    public int    VersionCount     { get; set; }

    // Scan
    public ScanStatus   ScanStatus        { get; set; } = ScanStatus.Pending;
    public DateTime?    ScanCompletedAt   { get; set; }
    public int?         ScanDurationMs    { get; set; }
    public List<string> ScanThreats       { get; set; } = new();
    public string?      ScanEngineVersion { get; set; }

    // Published logo tracking — true only when explicitly registered via the logo workflow
    public bool IsPublishedAsLogo { get; set; }

    // Soft delete
    public bool     IsDeleted  { get; set; }
    public DateTime? DeletedAt { get; set; }
    public Guid?    DeletedBy  { get; set; }

    // Compliance
    public DateTime? RetainUntil { get; set; }
    public DateTime? LegalHoldAt { get; set; }

    // Audit
    public DateTime CreatedAt  { get; set; }
    public Guid     CreatedBy  { get; set; }
    public DateTime UpdatedAt  { get; set; }
    public Guid     UpdatedBy  { get; set; }

    // Navigation
    public ICollection<DocumentVersion> Versions { get; set; } = new List<DocumentVersion>();
    public ICollection<DocumentAudit>   Audits   { get; set; } = new List<DocumentAudit>();

    public bool IsOnLegalHold => LegalHoldAt.HasValue;

    public static Document Create(
        Guid   tenantId,
        string productId,
        string referenceId,
        string referenceType,
        Guid   documentTypeId,
        string title,
        string? description,
        string mimeType,
        long   fileSizeBytes,
        string storageKey,
        string storageBucket,
        string? checksum,
        Guid   createdBy)
    {
        var now = DateTime.UtcNow;
        return new Document
        {
            Id             = Guid.NewGuid(),
            TenantId       = tenantId,
            ProductId      = productId,
            ReferenceId    = referenceId,
            ReferenceType  = referenceType,
            DocumentTypeId = documentTypeId,
            Title          = title,
            Description    = description,
            Status         = DocumentStatus.Draft,
            MimeType       = mimeType,
            FileSizeBytes  = fileSizeBytes,
            StorageKey     = storageKey,
            StorageBucket  = storageBucket,
            Checksum       = checksum,
            ScanStatus     = ScanStatus.Pending,
            IsDeleted      = false,
            CreatedAt      = now,
            CreatedBy      = createdBy,
            UpdatedAt      = now,
            UpdatedBy      = createdBy,
        };
    }
}
